File: Arrays and Vectors/LowerBound.cpp
#include<bits/stdc++.h>
using namespace std;

int lowerBound(vector<int> & arr, int target) {
    int n = arr.size();
    int lo = 0;
    int hi = n-1;
    int ans = -1;
    
    while(lo <= hi) {
        int mid = (lo + hi) / 2;
        if(arr[mid] <= target) {
            ans = arr[mid];
            lo = mid+1;
        }
        else {
            hi = mid-1;
        }
    }
    
    return ans;
}

int main() {
    vector<int> v = {-1, -1, 2, 3, 5};
    cout<<lowerBound(v, 4)<<endl;
}

File: Arrays and Vectors/LowerBound.js
const LowerBound = (arr, target) => {
  let n = arr.length;

  let lo = 0;
  let hi = n - 1;
  let lower_bound = -1;

  while (lo <= hi) {
    let mid = Math.floor((lo + hi) / 2);

    if (arr[mid] === target) {
      return arr[mid];
    } else if (target < arr[mid]) {
      hi = mid - 1;
    } else {
      lower_bound = arr[mid];
      lo = mid + 1;
    }
  }

  return lower_bound;
};

const arr = [1, 2, 4, 5, 6];
console.log(LowerBound(arr, 0));
